### ISSUE TITLE
runtime core issue while running runtime-nodejs

### ISSUE THREAD

**Issue by jboulmal** [reTHINK-project/dev-runtime-core#131]
![screenshot from 2016-11-03 12 45 12](https://cloud.githubusercontent.com/assets/15377466/19964673/6c05f7d0-a1c3-11e6-8585-bda30bbac9aa.png)
Recently, while testing storage-manager for runtime Node, we got the following issue :
I believe this issue comes from lasst updates on runtime-core.
```
5: return sandbox and register
--- Policy Engine ---
{ type: 'create',
  to: 'runtime://hysmart.rethink.ptinovacao.pt/4770/identity-gui',
  from: 'runtime://hysmart.rethink.ptinovacao.pt/4770/idm',
  body: { value: { identities: [], idps: [Object] } },
  id: 1 }
--- Policy Engine ---
{ id: 1,
  type: 'response',
  to: 'runtime://hysmart.rethink.ptinovacao.pt/4770/idm',
  from: 'runtime://hysmart.rethink.ptinovacao.pt/4770/identity-gui',
  body: { type: 'idp', value: 'google.com', code: 200 } }
selectedIdentity:  google.com
ReferenceError: crypto is not defined
    at null:8573:9
    at Promise.F (null:463:28)
    at Crypto.generateRSAKeyPair (null:8572:14)
    at null:9608:22
    at Promise.F (null:463:28)
    at IdentityModule.callGenerateMethods (null:9602:14)
    at null:9562:23
    at process._tickCallback (internal/process/next_tick.js:103:7)
(node:16515) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 4): Failed to obtain an identity
```

`` "runtime-core": "github:rethink-project/dev-runtime-core#develop",``
 `` "service-framework": "github:rethink-project/dev-service-framework#develop"``

**Comment by pchainho** [comment]
I understand the first release of nodejs runtime is pending on this one

@gamdias it seems to be related with IDM right?

**Comment by jboulmal** [comment]
yes @pchainho.Especially, in the dist/folder Runtime.js in `develop` branch:
```
 key: 'requestIdentityToGUI',
    value: function requestIdentityToGUI(identities, idps) {
      var _this = this;

      return new _promise2.default(function (resolve, reject) {

        //condition to check if the real GUI is deployed. If not, deploys a fake gui
        if (_this.guiDeployed === false) {

          var guiFake = new _GuiFake2.default(_this._guiURL, _this._messageBus);
          _this.guiFake = guiFake;
          _this.guiDeployed = true;
        }

        var message = { type: 'create', to: _this._guiURL, from: _this._idmURL,
          body: { value: { identities: identities, idps: idps } } };

        var id = _this._messageBus.postMessage(message);

        //add listener without timout
        _this._messageBus.addResponseListener(_this._idmURL, id, function (msg) {
          _this._messageBus.removeResponseListener(_this._idmURL, id);

          if (msg.body.code === 200) {
            var selectedIdentity = msg.body;

            console.log('selectedIdentity: ', selectedIdentity.value);
            resolve(selectedIdentity);
          } else {
            reject('error on requesting an identity to the GUI');
          }
        });
      });
    }
  }, {..................
```

**Comment by gamdias** [comment]
Hello @jboulmal, it is strange that the 'requestIdentityToGUI' method is being called, since it is inside of a verification, ex:
 try {
   if (window) {
       //NodeJS should NOT run this code
        requestIdentityToGUI();

    } catch (error) {
        //NodeJS should run this code
        createAFakeIdentity();
    }
}

Do you have any variable called window?

**Comment by jboulmal** [comment]
Hi @gamdias 
Yes I do define https://github.com/reTHINK-project/dev-runtime-nodejs/blob/node-storage-manager/src/RuntimeFactory.js#L65 for the storageManager shim to work.

**Comment by gamdias** [comment]
That is the variable that I used to differentiate the different runtimes (browser and nodeJS).
Can you change the name of the window variable?

If not, I'm open to new suggestions on how to differentiate the different runtimes :)

**Comment by pchainho** [comment]
The runtime type should be provided in a [RuntimeCapabilities json object to be provided by a RuntimeCapabilityManager component](https://github.com/reTHINK-project/specs/issues/13#issue-183757014), but this is not yet available. 

BTW, I still have to mention this in the updated version of the runtime spec :/

**Comment by jboulmal** [comment]
Differentiating environments at runtime is good feature, and everything after just follows the pre-set runtime environment. That would be great to have @pchainho! 
untill then @gamdias, defining `window` object as global is only way to make this [polyfill](https://github.com/axemclion/IndexedDBShim/blob/master/README.md#node) work in node.

**Comment by jboulmal** [comment]
What is the status from your side @gamdias ?

**Comment by gamdias** [comment]
I'm still waiting for the RuntimeCapabilities new features. No changes has been made, yet.

**Comment by pchainho** [comment]
You can try the [RuntimeCapabilities](https://github.com/reTHINK-project/dev-service-framework/tree/runtime-capabilities-manager/src/runtime-capabilities) new feature with [this toolkit branch](https://github.com/reTHINK-project/dev-hyperty-toolkit/tree/runtime-capabilities) created for testing purposes

**Comment by jboulmal** [comment]
We can verify now if `"windowSandbox": true/false,` or/and  `"node": true/false `
https://github.com/reTHINK-project/specs/tree/master/datamodel/core/hyperty-catalogue#hyperty-runtime-descriptor

**Comment by BernardoMG** [comment]
Runtime Core is already using capabilities, to verify the runtime used (ex: nodejs or browser).

The changes are committed in pull request https://github.com/reTHINK-project/dev-runtime-core/pull/138

**Comment by gamdias** [comment]
the changes in the identity module regarding the verification of the RuntimeCapabilities are up and running without any problem.

If there is any problem, please reopen this issue.

**Comment by jboulmal** [comment]
I did an npm update lately, and then I got this issue again!

**Comment by tiagolb** [comment]
This issue happened only in the runtime-capabilities branch of the dev-hyperty-toolkit, which was not up-to-date. Since this branch was merged into the develop branch, the error is corrected in this develop branch.
